### ISSUE TITLE
Settings strategy and introduction of new methods in place of the deprecated one

### ISSUE THREAD

**Issue by benjaoming** [learningequality/ka-lite#3757]
There are several ways of injecting your own settings:

1. By editing the source code **DISCOURAGED**. However, I think a lot of people might just do that.
1. By placing one's own local_settings.py in the system's `kalite` package **DEPRECATED**
1. Creating your own settings module and using the command line switch to point KA Lite and Django to it, for instance `kalite start --settings=my_module`. **NEW**
1. Same as above but using DJANGO_SETTINGS_MODULE to point django.conf to the settings module.

I would propose the following for 0.14:

1. To have `kalite.project.settings` created and functional
1. To make any references for `kalite.settings` raise a `DeprecationWarning`
1. Create command line option for most commonly used settings: `--listen-port`, `--debug`, `--content-folder`.
1. As far as possible, to get rid of `fle_utils.settingshelper`, not expecting this work to be conclusive yet.

Then, in 0.15, we can start cleaning up settings and do remaining work in case `fle_utils.settingshelper` wasn't fully dealt with.

One of the big issues lately has been how "PRODUCTION_PORT undefined blah blah" exceptions show up, but are actually caused by exceptions in settings modules that are magically imported by `fle_utils.settingshelper`. If this happens to end-users, we will have no way of helping them, because their exceptions will seem quite meaningless, as @EdDixon is currently experiencing with testing.

There's also an issue on writing documentation for this: #3611
And an issue for the over-all idea of making kalite a "normal" Django application: #2952

**Comment by MCGallaspy** [comment]
I'm most interested in points 1 and 3. By creating `kalite.project.settings` you essentially mean defining a bunch of default settings that can be imported, yes? What's preventing this right now, in `mycoolsettings.py`:
```python
import kalite.settings.base
MYCOOLSETTING = 42
```

**Comment by benjaoming** [comment]
@MCGallaspy nothing! The original idea was also to say (as I think the current deprecation warning message says) that you can do that.

But since settings are normally nested in the project scope, I'd like the move to consist of a refactor so we don't promote `from kalite.settings.base import *` but start out with the nicest possible practice, which I believe is to create a package for everything that's related to the project and not the app. For internal practices especially, having `settings` in `kalite` encourages that applications should interfere with settings, which they should never do, they should *read* settings, not try to access them and change them runtime.

Actually, IMO, the most full-blown version of this would be to completely remove the project from `kalite` and distribute a separate `kalite_project` package, but if we implement `kalite.project` in a proper way, we can feel free to do that later.. and I'm a little bit concerned that this is too radical because of how the kalitectl logic is actually build around addressing components of `kalite` as a project.

**Comment by MCGallaspy** [comment]
Let me make sure I'm understanding what you're saying. Ideally it should be the case that:

1. The project `kalite` has its own settings module which does not reference any particular app. (I think we currently violate this, right?)
2. Each app (`distributed`, `coachreports`, etc) has specific settings which don't modify the project's settings. (Again, we _do_ in fact modify project-wide settings within apps, right?)

And if a user wishes to define custom settings, you're suggesting:
1. We create a package (`kalite.project.settings`) which defines default settings. A user can import that module (`from kalite.project.settings import *`) and change default values. The module so created can be used with the `--settings` flag.
2. Commonly changed settings be given a cl flag; `--debug`, `--content-folder`, etc. (Though it's not clear to me how this should be implemented. Does you mean that `kalitectl` will create a new settings module from the cl options, before starting the server?)

Is that correct?

**Comment by benjaoming** [comment]
Yes, all correct!

Regarding app settings, it's important that apps follow this kind of pattern:

**kalite.some_app.settings**

```python
import os
from django.conf import settings as django_settings

# This one can be customized in project settings
SOME_SETTING = getattr(django_settings, 'MY_APP_SOME_SETTING', 'default_value')

# This one can be customized in project settings and even the ENV
SOME_OTHER_SETTING = getattr(
    os.environ,
    'MY_APP_SOME_OTHER_SETTING',
    getattr(django_settings, 'MY_APP_SOME_OTHER_SETTING', 'default_value')
)
```

But this pattern is ***illegal***:

**kalite.some_app.settings**

```python
from django.conf import settings

settings.INSTALLED_APPS += ['some_other_app']
```

...Because applications should never modify projects, then you get a mess regarding responsibility and possibility of inconsistency.

However, in the `kalite.project.settings` package, we should have modules reflecting specific domains, like if say `kalite.my_app` was special enough we would create a settings module to reflect its way of doing things:

**kalite.project.settings.my_app**

```python
from .base import set_application_dependency

set_application_dependency("other_app")
```

...and then, finally, in `kalite.project.settings.base`, we can tie everything together nicely like:

**kalite.project.settings.base**

```python
from .django_settings import *
from .my_app import *
# etc...
```

And if we want to make something special for Android, we can create `kalite.project.settings.deployments.android` and put:

**kalite.project.settings.deployments.android**

```python
from ..base import *
KALITE_CONTENT_HOME = 'Something else'  # Because Android doesn't use home directories
```

**Comment by jamalex** [comment]
Could you explain how settings from which other settings' values are "derived" can be handled in this system? For example, say I want to change `CONTENT_ROOT` -- where do I do that, noting that this then needs to affect the default values for `KHAN_CONTENT_PATH`, `ASSESSMENT_ITEM_DATABASE_PATH`, `ASSESSMENT_ITEM_VERSION_PATH`, `ASSESSMENT_ITEM_JSON_PATH`, etc?

As another example, we have many settings that depend on `DEBUG`. How does that work in the proposed approach?

If we don't have a clean solution to those cases, then I don't think we're ready to deprecate local_settings.py (which currently handles this problem very well). Also, we're already past code freeze, so our focus for 0.14 now needs to be on testing, not refactoring/new features, etc, otherwise we won't be shipping on time.

**Comment by aronasorman** [comment]
I'm all for huge refactorings that would reduce our technical debt, but I don't think this should come in to 0.14 anymore (as what the milestone indicates).

We basically have 2.1 weeks left 'til our scheduled release date, and right now none of our installers are working (except for pip). The central server is also in pretty bad shape due to the changes in how things are imported, + the changes in settings.

@jamalex mentioned in the dev standup meeting that since it isn't feasible yet to remove support for `local_settings.py` for 0.14, we should remove the deprecation warning for now. It might be good to restore that deprecation warning once we have this implemented, as that should give us time to test out this new settings system and work out the kinks. Because you never know what could go wrong :)

**Comment by benjaoming** [comment]
We know that in order to do this refactor, we need a deprecation warning. So this means not getting rid of the old ways until 0.16.

The deprecation warning means that we need to maintain support for fle_utils.settingshelper, that has cost me hours and hours of unproductive nonsense and will continue to cost users and devs problems in the future. Every now and then I get the impression that I'm not the only one who's spent time look at odd tracebacks.

Furthermore, we have a mission to accomplish regarding upgrading Django, doing py2+3, and having a South+django.db.migrations approach. If these features are wanted, I highly recommend getting rid of this problematic chuck for 0.15.

@jamalex 
> Could you explain how settings from which other settings' values are "derived" can be handled in this system?

It's tricky, but say you have DEBUG and you need this to be inject-able from an outside env, what you can do is this:

**kalite.project.settings.base**

```python
import os
DEBUG = getattr(os.environ, 'DJANGO_DEBUG', False)
```

So what's the purpose of not having some `from users_module import *` and then use `hasattr` like the old approach? Well, basically this behaviour is ill-defined.

This seems to fit quite well for the `DEBUG` setting, because it would commonly be needed as a `--debug` flag anyways.

For the part of the paths, I think it's already a problematic approach to say "if the user sets this setting X to 1, then setting Y should be 1+2".. and to have all that happen in the settings stack. It's very implicit.

If we have a path that by default depends on CONTENT_ROOT, then make it so in the applications that are using it contain that behaviour, and not the settings stack:

**kalite.some_app.settings**

```python
import os
from django.conf import settings as django_settings

MY_OTHER_PATH = getattr(
    django_settings,
    'MY_OTHER_PATH',
    os.path.join(django_settings.CONTENT_ROOT, 'sub-folder')
)
```

**Comment by jamalex** [comment]
> `DEBUG = getattr(os.environ, 'DJANGO_DEBUG', False)`

Ok, yes, environment variables would make sense in that case. And I agree it would be nice in a purist sense to avoid too many derived settings, but it's often tricky in practice (at least until we've restructured things around a "core" / "plugins" approach). And in particular, this isn't something we're going to have ready for 0.14, so I'd like to see a much more constrained plan for what's going to happen prior to this (very imminent) release.

**Comment by benjaoming** [comment]
But my suggestion wasn't to target what's outlined in the examples.

For 0.14, it was to expose an API to users/devs so we can get old methods deprecated. It says so pretty clearly in the original ticket.

**Comment by rtibbles** [comment]
Assuming this has been fixed in #3809.

**Comment by benjaoming** [comment]
Well. all this is underway. I guess #4054 covers this issue, so might as well close it again :)
